Skip tests if NotImplementedError is thrown.

jzohrab opened the pull request:
Currently, Napalm tests have many NotImplementedErrors due to devices not having all features.  As a dev and longtime TDD guy, I'm accustomed to running my tests, and having them either pass or fail.  I find that false negatives clutter my test runs and interrupt my dev "flow", eg:

```
(venv) ~/Documents/Projects/napalm-nxos$ python -m unittest test.unit.TestNXOSDriver
.E.EE..E......EEE..E....
... lots of (intentional) error noise ...
```

This PR checks for NotImplementedErrors for each method call, and if that error is raised, it skips the test.  Here's the same test run after the patch is applied:

```
(venv) ~/Documents/Projects/napalm-nxos$ python -m unittest test.unit.TestNXOSDriver
.s.ss..s......sss..s....
----------------------------------------------------------------------
Ran 24 tests in 0.046s

OK (skipped=8)
```

You can see what has been skipped with the verbose flag:

```
(venv) ~/Documents/Projects/napalm-nxos$ python -m unittest --verbose test.unit.TestNXOSDriver 2>&1 | grep skipped
test_get_bgp_config (test.unit.TestNXOSDriver.TestGetterNXOSDriver) ... skipped ''
test_get_bgp_neighbors_detail (test.unit.TestNXOSDriver.TestGetterNXOSDriver) ... skipped ''
test_get_environment (test.unit.TestNXOSDriver.TestGetterNXOSDriver) ... skipped ''
test_get_interfaces_counters (test.unit.TestNXOSDriver.TestGetterNXOSDriver) ... skipped ''
test_get_probes_config (test.unit.TestNXOSDriver.TestGetterNXOSDriver) ... skipped ''
test_get_probes_results (test.unit.TestNXOSDriver.TestGetterNXOSDriver) ... skipped ''
test_get_route_to (test.unit.TestNXOSDriver.TestGetterNXOSDriver) ... skipped ''
test_ping (test.unit.TestNXOSDriver.TestGetterNXOSDriver) ... skipped ''
```

The Python language designers actually discussed special handling for NotImplementedError in http://bugs.python.org/issue1466, but that feature isn't implemented.  I don't really want to subclass the test runner, this change I propose is much simpler.

I think this would simplify testing for all projects as a whole.  Discussion welcome!

mirceaulinic commented on napalm_base/test/test_NapalmTestFramework.py:
```diff
@@ -0,0 +1,14 @@
+from napalm_base.test.base import TestConfigNetworkDriver, TestGettersNetworkDriver
+
+
+class TestGetterNetworkDriver_passes(unittest.TestCase, TestGettersNetworkDriver):
```
Where do you expect this to be used and how?

jzohrab replied:
This test class was mostly a sanity check for me only to ensure that the skip covered all cases.  Plus I like to add tests for pretty much everything to ensure I have a good understanding of how things fit together.

I really don't expect this test to be run much, maybe add it to Travis, but it's kind of silly on its own. :-)

dbarrosop replied:
What's the benefit of this? Getters are mocked already, so you can just run the test without this `FakeThing` as there is a fake class already on getter test.

jzohrab replied:
I wasn't aware.  I read through https://github.com/napalm-automation/napalm-base/blob/develop/napalm_base/test/base.py and couldn't see an existing fake. Since the TestGettersNetworkDriver doesn't inherit from TestCase, it's not run as a test, so I created this test by copying the outline from one of the implementations and reducing it to this.  Am I looking in the wrong spot?

If there is already a test somewhere that calls all the methods in TestGettersNetworkDriver with a fake object then of course this is redundant.  If not, including this test would be good for completeness -- e.g., if a new method is added to the base driver, and its test doesn't handle NotImplementedError, then this test would catch that.

Really, it's a test of the tests, which is sometimes good for project stability.  It's not a problem for me if you would prefer to remove it or disable it, I'm presenting it as an option only.

dbarrosop replied:
> if a new method is added to the base driver, and its test doesn't handle NotImplementedError, then this test would catch that

I was trying to understand the motivation and that one seems like a good one for : ) Could you enable it in `.travis.yml`?

jzohrab replied:
Before I do that, is this the right location and name for this file?  It's in `napalm_base/test/test_NapalmTestFramework.py`.  I could move this to be part of the `base.py` file, which might be a better location for it, and would be right next to the TestGetter class itself.

dbarrosop replied:
Use `/napalm-base/test/unit/TestNapalmTestFramework.py` and name the class `TestSkipNotImplemented`.

mirceaulinic commented:
The tests are mainly executed by Travis, but for sure this is very useful when running them locally!

jzohrab commented:
I prefer having a test suite that I can run on my desktop, as much as possible.  Should be doable with fakes, even if a dev doesn't have a full set of devices to test against.

jzohrab commented:
Incorporated notes from @dbarrosop, added to Travis.  Runs OK locally:

```
$ python -m unittest test.unit.TestNapalmTestFramework
sssssssssssssssssssss
...[snip]...
OK (skipped=21)
```

jzohrab commented:
Travis failed with `TestGettersNetworkDriver instance has no attribute 'device'`.  I'm running the tests successfully with unittest, I guess there is something different in the setup.  Investigating.

At the end of this will likely rebase and squash, remove the commit history noise.

jzohrab commented:
Rebased, as commits were messy due to Travis failures.  The current code works with both unittest and nose:

```
$ python -m unittest test.unit.TestNapalmTestFramework
sssssssssssssssssssss
...
OK (skipped=21)
$ nosetests ./test/unit/TestNapalmTestFramework.py
SSSSSSSSSSSSSSSSSSSSS
...
OK (SKIP=21)
```

jzohrab commented:
Nice, passed.

@dbarrosop, this PR may also simplify the .travis.yml in all the projects.  eg, in the Nexus project, rather than having to spell out all the tests you want to include in its [.travis.yml](https://github.com/napalm-automation/napalm-nxos/blob/develop/.travis.yml), you should be able to specify the full test class.

mirceaulinic commented on .travis.yml:
```diff
@@ -20,3 +20,4 @@ deploy:
 script:
   - nosetests ./test/unit/TestGetNetworkDriver.py
   - nosetests ./test/unit/TestHelpers.py
+  - nosetests ./test/unit/TestNapalmTestFramework.py
```
```LF``` perhaps?

mirceaulinic replied:
Anyway maybe this test should be removed?

mirceaulinic commented on test/unit/TestNapalmTestFramework.py:
```diff
@@ -0,0 +1,17 @@
+# nose finds it and tries to run the tests for it, which yields
+# errors.
+
+class TestSkipNotImplemented(unittest.TestCase, ntb.TestGettersNetworkDriver):
```
I assume this will be removed from the PR, am I right?

jzohrab replied:
@mirceaulinic - I was thinking it would stay, per the discussion with David in https://github.com/napalm-automation/napalm-base/pull/41#discussion-diff-67585869.  If a new method is added to the base driver, and its test doesn't handle NotImplementedError, then this test would catch that.

Ideally, the NotImplementedError will be used to simplify the .travis.yml files in each project, and so it would need to continue to work going forward.

mirceaulinic replied:
Right, did not see this discussion, sorry.

jzohrab replied:
NP, clearly the comment "Ensure that any tests are skipped if not implemented" was insufficient, but that was the intent ... I could add more notes if you think that would improve clarity.  Say "yes please" and I'll do it.

mirceaulinic replied:
This is fine, it is clear enough. It was just my bad I did not read carefully.

mirceaulinic commented:
🏆
